src/vm/peimage.cpp

```diff
@@ -1104,7 +1104,10 @@ void PEImage::Load()
 
     SimpleWriteLockHolder lock(m_pLayoutLock);
 
-    _ASSERTE(m_pLayouts[IMAGE_LOADED] == NULL);
+    if (m_pLayouts[IMAGE_LOADED] != NULL)
+    {
+        return;
+    }
```

Comment:
The `HasLoadedLayout` called above is the following:
```inline BOOL PEImage::HasLoadedLayout()                                                                                 
{                                                                                                                      
    LIMITED_METHOD_CONTRACT;                                                                                           
    SUPPORTS_DAC;                                                                                                      
    return m_pLayouts[IMAGE_LOADED]!=NULL;                                                                             
}
```

It actually performs the same check.

I think there is a multithreading race condition for value of the `m_pLayouts[IMAGE_LOADED]`.

So, moving the lock acquisition above the `HasLoadedLayout` check should also fix the issue.

This implementations is also correct, and can be a performance optimization, because it doesn't perform lock acquisition if layout is already loaded.

I propose to implement the new `if (m_pLayouts[IMAGE_LOADED] != NULL)` with `if (HasLoadedLayout())`, and to add a comment which describes that this is a performance optimization, like the following:

```
void PEImage::Load()
{
    STANDARD_VM_CONTRACT;

    if (HasLoadedLayout())
    {
        _ASSERTE(GetLoadedLayout()->IsMapped()||GetLoadedLayout()->IsILOnly());
        return;
    }

    SimpleWriteLockHolder lock(m_pLayoutLock);

    // The condition could be changed before lock was acquired, so repeating the check
    // This is a performance optimization to avoid acquiring lock in case layout is already loaded
    if (HasLoadedLayout())
    {
       return;
    }

   _ASSERTE(m_pLayouts[IMAGE_LOADED] == NULL);
```

or to move the lock acquisition above the first check.